### ISSUE TITLE
Multi-root workspaces

### ISSUE THREAD

**Issue by MarkZ3** [theia-ide/theia#1660]
It would be great to be able to open many different folders in the same workspace, regardless of where they are located on the file system. Then the workspace settings could be located separately of the root folders (and not pollute the Git repos for example). There is already such support for this in VSCode and the LSP, see
https://code.visualstudio.com/docs/editor/multi-root-workspaces
https://github.com/Microsoft/language-server-protocol/issues/281

**Comment by elaihau** [comment]
These days I have been thinking about what needs to be done to preferences to better support having multiple roots. 
This afternoon we had a discussion. Please see below for the problem we are trying to resolve and the potential solution and issues we would have.
Any ideas & suggestions are greatly appreciated! @akosyakov @svenefftinge 

**Problem:**

With the partially implemented prototype of multi-root workspace support, `WorkspacePreferenceProvider` reads preferences only from the `activeRoot` folder and apply them across the entire workspace, causing issues where the prefrerences in `non-active roots` are ignored. See code snippet below for details.

At the moment, `activeRoot` in my code is the first root added to the workspace. 
```ts
@injectable()
export class WorkspacePreferenceProvider extends AbstractResourcePreferenceProvider {

    @inject(WorkspaceService)
    protected readonly workspaceService: WorkspaceService;

    protected async getUri(): Promise<URI> {
        const activeRoot = await this.workspaceService.activeRoot;
        if (activeRoot) {
            const rootUri = new URI(activeRoot.uri);
            return rootUri.resolve('.theia').resolve('settings.json');
        }
        return new Promise<URI>(() => { });
    }
}
```

**Plan**

In theia, preferences can be set at the following 3 levels
* Root/Folder preferences
* User preferences
* Preference default values
where the rule of `Root/Folder preferences > User preferences > Preference default values` applies to resolve any conflicts.

The plan is 
1. adding `Workspace preference`, which has the highest priority, on top of them:
P0: Workspace preference
P1: Root/Folder preferences
P2: User preferences
P3: Preference default values
In other words, the `Workspace preference` would apply to all folders in the multi-root workspace. 
2. When users go from a single root instance to multi-root workspace, theia would add `Root/Folder preferences` of the first folder to the new workspace preferences.
3. the workspace preferences would be stored in the workspace config file.


**Potential issues**

Some editor-related preferences, e.g., useTabs & tabSize, are usually more language-specific. Using the `Root/Folder preferences` of the first folder could be problematic, especially when users have projects written in diffrerent languages added to the same workspace. The longer term plan is having the per-language editor-related preferences.

**Comment by gorkem** [comment]
@vparfonov I think this affects https://github.com/eclipse/che/issues/9336

**Comment by lmcbout** [comment]
Testing Multi-root raises some issues:
- [ ] Single folder workspace uses teh following structure in ".theia/recentworkspace {"recentRoots": 
      ["file:///..."]}. Can we use in Multi-root structure "recentsRoots" instead of "roots", it would make the data 
      in recentworkspace compatible with other versions of Theia.
- [ ] Multi-root does not start with a workspace argument:  yarn start ../../../workspace/work1
- [ ] Unabe to distinguish which folder is the active one, may need some kind of indication
- [ ] Search: search one or all workspace folder ?
- [ ] Can we save a multi-workspace with a workspace name ? If I create a few workspace with different folder, I would like to be able to re-open that workspace without having to re-open each folder from that multi-root workspace.
- [ ] With a workspace name, can we display the name of that workspace somewhere, so it is easy to the end-ser to know the current workspace he is working on.
- [ ] Not to polute the ".theia" forder, we should have a "workspace" folder to store the created user workspace
- [ ] When selecting Open workspace, should we propose the list of already created workspace instead of the list of folder. This could be acheive if we get the name from the .theia/workspace/workspace...json"
- [ ]  How Can I change the active root so when opening a terminal, it opened in the proper folder!
- [ ] Drag & Drop a folder to another location  seems to make a copy to the new location, but it looks like the primary folder gets empty , the primary folder does not get removed from old location ?
- [ ] After  a Drag & Drop, now move the folder back to its original location => Files are moved, then try to open any file from the workspace generates an error and no files opened.
    [2018-05-30T12:45:47.011Z]  WARN: Theia/11736 on elxa36mbq32: Widget was activated, but did 
    not accept focus: theia-gitContainer [] (logger=root)

    [2018-05-30T12:46:51.187Z] ERROR: Theia/11736 on elxa36mbq32: (logger=root)
       [nsfw-watcher: 11810] 
       /home/lmcbout/theia/theia/node_modules/nsfw/lib/src/index.js:62
          throw nsfwError;
          ^
       Service shutdown unexpectedly

    [2018-05-30T12:47:35.677Z] ERROR: Theia/11736 on elxa36mbq32: Request watchFileChanges 
    failed with error: Connection is closed. [ 'Error: Connection is closed.\n    at new ConnectionError 
    (/home/lmcbout/theia/theia/node_modules/vscode-jsonrpc/lib/main.js:138:28)\n  ....

**Comment by tsmaeder** [comment]
I don't think trying to be "a bit pregnant" will get us very far here. We should focus on the concept of "workspace" vs. "workspace folder" and solve this problem first. What I would propose is a refactoring that solves the current behaviour of single workspace folder as a special case of multiple folders. Once we have this, we can start surfacing the capability in the UI.
If a workspace is a thing, it needs a place to store data that is independent of the projects it contains. I don't think the concept of "active workspace folder" introduced in https://github.com/theia-ide/theia/pull/2117 is advancing us in the right direction: it will introduce another intermediate version of theia metadata and lead to odd effects (like it would matter whether you add project a or project b first to the workspace when it comes go workspace preferences). 
If this is a task that is too hard to do in one go, let us refactor the system to a state where the task gets easier first.

**Comment by elaihau** [comment]
most features have been finished and merged. i believe this one can be closed.
